pkg/sr/serde.go

```diff
@@ -221,17 +284,17 @@ func tserdeMapClone(m map[int]tserde, at int, index []int) map[int]tserde {
 
 // Encode encodes a value according to the schema registry wire format and
 // returns it. If EncodeFn was not used, this returns ErrNotRegistered.
-func (s *Serde) Encode(v any) ([]byte, error) {
-	return s.AppendEncode(nil, v)
+func (s *Serde) Encode(v any, opts ...EncodingOpt) ([]byte, error) {
+	return s.AppendEncode(nil, v, opts...)
 }
 
 // AppendEncode appends an encoded value to b according to the schema registry
 // wire format and returns it. If EncodeFn was not used, this returns
 // ErrNotRegistered.
-func (s *Serde) AppendEncode(b []byte, v any) ([]byte, error) {
-	t, ok := s.loadTypes()[reflect.TypeOf(v)]
-	if !ok || (t.encode == nil && t.appendEncode == nil) {
-		return b, ErrNotRegistered
+func (s *Serde) AppendEncode(b []byte, v any, opts ...EncodingOpt) ([]byte, error) {
```

Comment:
What's the point of having options at the point of encoding?
Same thought on decoding.
I'd like to avoid options on these functions if possible -- would it be possible to have a block of `Dyn(Append)?Encode` functions?

Reply:
Hey @twmb! First of all, sorry, this work fell through the cracks. I'll try to find some time over the weekend and clean up the code. IIUC you're proposing to have `DynEncode` and `DynAppendEncode` that take `EncodingOpt` and leave the existing encode functions as they are, correct? I'll try it out.

Reply:
As promised, I added the changes. It simplified the code quite a bit - no more `ID` option, no more special treatment of any option, no more options passed to encode and decode. Instead we now have `DynEncode` and related functions, which expect parameters `id` and `index` to determine the correct `tserde`.

I'm looking forward to hear your feedback 😉

Reply:
Random question now that I'm seeing this PR:

Is there any reason to require registering the types at all and then using DynEncode, vs. having top-level, non Serde method functions,

```go
func Encode(id int, index []int, v any, enc func(any) ([]byte, error)) ([]byte, error)
func AppendEncode(b []byte, id int, index []int, v any, enc (func([]byte, any) ([]byte, error)) ([]byte, error)
```
?

I'm wondering: what use case do you have that allows you to know the id and index you want to encode, but not know the encode function?

Reply:
Oh well, this was another long delay, sorry about that. You know, it's the standard apology, being busy. I'll try to be more responsive now and help you resolve this PR.

I merged the changes from master and resolved the conflicts introduced in the PR you mentioned.

> what use case do you have that allows you to know the id and index you want to encode, but not know the encode function?

It's not that I don't know the encode function, I do know it actually. It's that `Serde` also appends the header and I don't want to duplicate that logic outside of `Serde`. Another reason to register the type is that the next time the type gets decoded in a different path of the code, we already have a `Serde` that knows how to decode it, no need to fetch the schema.

Reply:
No worries on the delay, I'm just as slow in the past few months due to games and traveling and holidays and work :)

I'm not sure I still understand but I'm sure I would if I see what you're trying to do -- do you have an example of how you intend to use the dyn functions that can't be served with the current API?

Reply:
Here's [the line](https://github.com/ConduitIO/conduit/blob/dbe5b8618ed92adb10650496fb73dee81146c39b/pkg/processor/schemaregistry/encoder.go#L53) where we call `Serde.Encode` and provide an ID of the schema to use. (For completeness, [here's](https://github.com/ConduitIO/conduit/blob/92d9681ed66abd17d5bef2b10ec434ee89fba818/pkg/processor/schemaregistry/decoder.go#L42) `Serde.Decode`).

I re-read your suggestion a few comments above and I think that having top-level, non Serde functions for `Encode` and `AppendEncode` that let you specify the ID and the encode func could work. The calls wouldn't go through the `Serde` type, but I realized we don't reuse the same object in the decode and encode paths of the code, so no harm done. And the solution feels cleaner than having `DynEncode`.

I'll see if I can get some time in the next few weeks to work this into the PR.